code/renderer/tr_terrain.c

```diff
@@ -1340,10 +1340,137 @@ void R_AddTerrainSurfaces()
 	}
 }
 
+/**
+ * Calculates the height of a terrain polygon by interpolating the heights of its vertices.
+ * The function finds the triangle containing the given point (x, y) within the terrain patch,
+ * and then computes the height at each vertex of the polygon using barycentric interpolation.
+ * Finally, it updates the z-coordinate of each vertex to store the calculated height.
+ * Returns qtrue if the triangle was found and the height could be calculated and saved into its vertices.
+ *
+ * ToDo: try to use the Vector*(?) macros for the math operations below from q_shared.h and q_math.h, not sure which ones would be a good fit
+ */
 qboolean R_TerrainHeightForPoly(cTerraPatchUnpacked_t* pPatch, polyVert_t* pVerts, int nVerts)
 {
-	// FIXME: unimplemented
-	return qfalse;
+	float x0 = 0;
+	float y0 = 0;
+	float x1 = 0;
+	float y1 = 0;
+	float x2 = 0;
+	float y2 = 0;
+	float fKx[3] = { 0 }, fKy[3] = { 0 }, fKz[3] = { 0 }, fArea[3] = { 0 };
+	float fAreaTotal = 0;
+
+	float x = pVerts->xyz[0];
+	float y = pVerts->xyz[1];
+
+	// Calculate the average of the x and y coordinates of all vertices
+	if (nVerts > 1)
+	{
+		for (int i = 1; i < nVerts; i++)
+		{
+			x += pVerts[i].xyz[0];
+			y += pVerts[i].xyz[1];
+		}
+
+		// Use the averaged values from this point onwards
+		x = x / nVerts;
+		y = y / nVerts;
+	}
+
+	// Get the first valid triangle in the patch
+	terraInt iTri = pPatch->drawinfo.iTriHead;
+	if (!iTri)
+	{
+		// Original assert message:
+		// "R_TerrainHeightForPoly: point (%f %f) not in patch (%f %f %f)\n", x, y, pPatch->x0, pPatch->y0, pPatch->z0
+		assert(!pPatch->drawinfo.iTriHead);
```

Comment:
For the message, the **va** function can be used like this:

```cpp
assert(!pPatch->drawinfo.iTriHead && va("...message..."))
```

Reply:
Good point, I was contemplating whether to add that but it felt hacky and haven't seen asserts being done like this anywhere else in the codebase. Either way, I'm adding it now, thanks for the tip!